File: js/problem20.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    if (s.length % 2 === 1) {
        return false;
    }
    
    let stack = [];
    
    for (let i = 0; i < s.length; i++) {
        if (s[i] === '(') {
            stack.push('(');
        } else if (s[i] === '{') {
            stack.push('{');
        } else if (s[i] === '[') {
            stack.push('[');
        } else if (s[i] === ')') {
            if ((stack.length === 0) || (stack[stack.length - 1] !== '(')) {
                return false;
            } else {
                stack.pop();
            }
        } else if (s[i] === '}') {
            if ((stack.length === 0) || (stack[stack.length - 1] !== '{')) {
                return false;
            } else {
                stack.pop();
            }
        } else if (s[i] === ']') {
            if ((stack.length === 0) || (stack[stack.length - 1] !== '[')) {
                return false;
            } else {
                stack.pop();
            }
        }
    }
    
    return stack.length === 0;
};

File: leetcodeJava/src/main/java/Problem20.java
// 20. Valid Parentheses

class Solution {
    public boolean isValid(String s) {
        if (s.length() == 0) {
            return true;
        } else if (s.length() % 2 == 1) {
            return false;
        }
        
        Stack<Character> stack = new Stack<Character>();
        for (int i = 0; i < s.length(); i++) {
            char current = s.charAt(i);
            if (current == '(' || current == '{' || current == '[') {
                stack.push(current);
            } else {
                if (current == ')') {
                    if (stack.empty() || stack.peek() != '(') {
                        return false;
                    }
                    stack.pop();
                } else if (current == '}') {
                    if (stack.empty() || stack.peek() != '{') {
                        return false;
                    }
                    stack.pop();
                } else {
                    if (stack.empty() || stack.peek() != '[') {
                        return false;
                    }
                    stack.pop();
                }
            }
        }
        
        return stack.empty();
    }
}